### PR TITLE
Emit an error when UnknownDeviceError occurs during VoIP operations

### PR DESCRIPTION
**Pull request by lukebarnard1**

### CODE REVIEW

**Comment by ara4n** [comment]
i think we may need this in a lot more places than just `m.call.invite` - e.g. attaachments, m.call.answer, m.call.hangup etc.  is there a more generic place? or just a matter of putting in lots of bubbles?

**Comment by lukebarnard1** [comment]
The obvious generic place is 
```js
const sendEvent = function(self, eventType, content) {
    return self.client.sendEvent(self.roomId, eventType, content);
};
```

**Comment by richvdh** [comment]
Looks fine. Could you document the new event type and `error` at the bottom, in a similar way to https://github.com/matrix-org/matrix-js-sdk/blob/e571271673fea843c1f69c9ca456bff86814adcf/src/client.js#L3006, so that we remember what the difference is? I'm particularly interested in recording the behaviour of MatrixCall when it encounters a `send_event_error` condition, and what your options are as an app when you get such an event.

File: src/client.js [hunk 1]
```diff
@@ -3020,6 +3020,25 @@ module.exports.CRYPTO_ENABLED = CRYPTO_ENABLED;
...
+ * does a hangup when it encounters a normal `error`, which in turn could lead to an
+ * UnknownDeviceError.)
+ * <p>
+ * To deal with an UnknownDeviceError when trying to send events, the application should let
+ * users know that there are new devices in the encrypted room (into which the event was
+ * sent) and give the user the options to resend unsent events or cancel them.
+ *
+ * @event module:client~MatrixClient#"send_event_error"
+ * @param {Error} err The error caught from calling client.sendEvent in call.js.
+ * matrixClient.on("send_event_error", function(err){
```
Comment (richvdh) [inline src/client.js]: example needs `@example`

File: src/client.js [hunk 2]
```diff
@@ -3020,6 +3020,25 @@ module.exports.CRYPTO_ENABLED = CRYPTO_ENABLED;
  */
 
 /**
+ * Fires whenever an error occurs when call.js sendEvent encounters an error.
+ * <p>
+ * This is required to allow UnknownDeviceError to propagate from client.sendEvent, whilst
```
Comment (richvdh) [inline src/client.js]: it's not specific to unknowndeviceerror though? it's about all event-transmission errors?

File: src/client.js [hunk 3]
```diff
@@ -3020,6 +3020,25 @@ module.exports.CRYPTO_ENABLED = CRYPTO_ENABLED;
...
+ * differentiating it from being a normal `error` emission. (This is because call.js
+ * does a hangup when it encounters a normal `error`, which in turn could lead to an
+ * UnknownDeviceError.)
+ * <p>
+ * To deal with an UnknownDeviceError when trying to send events, the application should let
+ * users know that there are new devices in the encrypted room (into which the event was
+ * sent) and give the user the options to resend unsent events or cancel them.
+ *
```
Comment (richvdh) [inline src/client.js]: worth noting what the MatrixCall does. Does it hang up? does it carry on regardless? does it block until you resend the event?

File: src/client.js [hunk 4]
```diff
@@ -3020,6 +3020,25 @@ module.exports.CRYPTO_ENABLED = CRYPTO_ENABLED;
...
+ * Fires whenever an error occurs when call.js sendEvent encounters an error.
+ * <p>
+ * This is required to allow UnknownDeviceError to propagate from client.sendEvent, whilst
+ * differentiating it from being a normal `error` emission. (This is because call.js
+ * does a hangup when it encounters a normal `error`, which in turn could lead to an
+ * UnknownDeviceError.)
+ * <p>
+ * To deal with an UnknownDeviceError when trying to send events, the application should let
+ * users know that there are new devices in the encrypted room (into which the event was
+ * sent) and give the user the options to resend unsent events or cancel them.
```
Comment (richvdh) [inline src/client.js]: how does the app resend/cancel unsent events?

File: src/client.js [hunk 5]
```diff
@@ -3020,6 +3020,25 @@ module.exports.CRYPTO_ENABLED = CRYPTO_ENABLED;
  */
 
 /**
+ * Fires whenever an error occurs when call.js sendEvent encounters an error.
```
Comment (richvdh) [inline src/client.js]: `sendEvent` is an implementation detail. "Fires when the MatrixCall encounters an error when sending a Matrix event".

File: src/client.js (hunk 2 above)
Comment (richvdh) [inline src/client.js]: I think this whole paragraph is unnecessary tbh.

File: src/client.js [hunk 6]
```diff
@@ -3020,6 +3020,25 @@ module.exports.CRYPTO_ENABLED = CRYPTO_ENABLED;
...
+ * This is required to allow UnknownDeviceError to propagate from client.sendEvent, whilst
+ * differentiating it from being a normal `error` emission. (This is because call.js
+ * does a hangup when it encounters a normal `error`, which in turn could lead to an
+ * UnknownDeviceError.)
+ * <p>
+ * To deal with an UnknownDeviceError when trying to send events, the application should let
+ * users know that there are new devices in the encrypted room (into which the event was
+ * sent) and give the user the options to resend unsent events or cancel them.
+ *
+ * @event module:client~MatrixClient#"send_event_error"
```
Comment (richvdh) [inline src/client.js]: ought to be associated with the MatrixCall, not the MatrixClient.

**Pull request merged**